### PR TITLE
Add support for transient metadata to IngestDocument

### LINKED ISSUE #15036: [Ingest] Support for ingest transient metadata
We should add support for transient metadata fields that get added by ingest to each document that gets processed, which can later be referred to by any processor. For instance the ingest timestamp would be a good one to be available all the time; it should be possible to add a field through the add processor and set it to that value. Same for information about the error when a processor fails, so that the on failure processor can grab it and write it to the desired field using the add processor.

Yet to be defined is the syntax to refer to these fields, we have to be careful and try to prevent confusion with elasticsearch metadata fields.

Relates to #14990 .

### PR DESCRIPTION
**Pull request by javanna**
IngestDocument now holds an additional map of transient metadata. The only field that gets added automatically is `timestamp`, which contains the timestamp of ingestion in ISO8601 format. In the future it will be possible to eventually add or modify these fields, which will not get indexed, but they will be available via templates to all of the processors.

Transient metadata will be returned by the simulate api, although they will never get indexed. Moved `WriteableIngestDocument` to the simulate package as it's only used by simulate and it's now modelled for that specific usecase.

 Also taken the chance to remove one IngestDocument constructor used only for testing (accepting only a subset of es metadata fields). While doing that introduced some more randomizations to some existing processor tests.

TODO: 
1) define name for the timestamp transient field and the ancestor object for the transient metadata fields in the simulate api (will be aligned with how we decide to expose those fields through templates)
2) expose failure as a transient metadata field when something breaks, maybe should be done after the new on_failure handling is in

Closes #15036

### CODE REVIEW

File: plugins/ingest/src/main/java/org/elasticsearch/ingest/IngestDocument.java [hunk 1]
```diff
@@ -337,12 +353,28 @@ private void setFieldValue(String path, Object value, boolean append) {
...
 
-    public String getMetadata(MetaData metaData) {
-        return this.metaData.get(metaData.getFieldName());
+    public String getEsMetadata(MetaData esMetadata) {
+        return this.esMetadata.get(esMetadata.getFieldName());
     }
 
-    public void setMetaData(MetaData metaData, String value) {
-        this.metaData.put(metaData.getFieldName(), value);
+    public Map<String, String> getEsMetadata() {
```
Comment (uboness) [inline plugins/ingest/src/main/java/org/elasticsearch/ingest/IngestDocument.java]: why do we differentiate between es vs. ingest metadata? why not just have metadata and then we can use namespaces for differentiating between different types of fields (e.g. `es.key` vs. `ingest.key`)?

Comment (martijnvg) [inline plugins/ingest/src/main/java/org/elasticsearch/ingest/IngestDocument.java]: I think we should have two maps. One map holding es metadata and source, and the other map holding the ingest metadata. This is going to fit nicely with the way we're going to refer to source and metadata fields. (just `_id` for id of the doc, `field1` to get a field name field1 from the source. Since meta fields can't be used in the source this should work out) The `IngestDocument.MetaData` enum can used to detect if a top level entry is either part of the source or metadata.

**Comment by martijnvg** [comment]
> define name for the timestamp transient field and the ancestor object for the transient metadata fields in the simulate api (will be aligned with how we decide to expose those fields through templates)

I think the `timestamp` is good name. They way we're going to refer to it will be `_ingest.timestamp` so that looks good to.

> expose failure as a transient metadata field when something breaks, maybe should be done after the new on_failure handling is in

I think this should be done as part of the on_failure task #14548.

Beyond the comment about the different maps in IngestDocument this looks good.

File: plugins/ingest/src/main/java/org/elasticsearch/ingest/IngestDocument.java (hunk 1 above)
Comment (javanna) [inline plugins/ingest/src/main/java/org/elasticsearch/ingest/IngestDocument.java]: I made the change suggested by Martijn. I think ingest metadata does need to be in a different map to prevent field name conflicts, unless we change approach regarding how we store _source and metadata fields in a single map, but that looks good at the moment I think.

**Comment by javanna** [comment]
@martijnvg I pushed new commits

File: plugins/ingest/src/main/java/org/elasticsearch/ingest/IngestDocument.java [hunk 2]
```diff
@@ -449,7 +451,32 @@ public static MetaData fromString(String value) {
                     throw new IllegalArgumentException("no valid metadata field name [" + value + "]");
             }
         }
-
     }
 
+    private class FieldPath {
```
Comment (martijnvg) [inline plugins/ingest/src/main/java/org/elasticsearch/ingest/IngestDocument.java]: maybe make this class static?

Comment (martijnvg) [inline plugins/ingest/src/main/java/org/elasticsearch/ingest/IngestDocument.java]: never mind... it can't

Comment (javanna) [inline plugins/ingest/src/main/java/org/elasticsearch/ingest/IngestDocument.java]: Shall I simply pass in sourceAndMetadata and ingestMetadata maps to the constructor?

File: plugins/ingest/src/main/java/org/elasticsearch/ingest/IngestDocument.java [hunk 3]
```diff
@@ -449,7 +451,32 @@ public static MetaData fromString(String value) {
...
+            if (newPath.startsWith("_source.")) {
+                newPath = newPath.substring(8, path.length());
+            }
+
+            this.pathElements = Strings.splitStringToArray(newPath, '.');
+            if (pathElements.length == 0) {
+                throw new IllegalArgumentException("path [" + path + "] is not valid");
+            }
+        }
+    }
```
Comment (martijnvg) [inline plugins/ingest/src/main/java/org/elasticsearch/ingest/IngestDocument.java]: I like this class, nice way of encapsulating the field path logic.

File: plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineExecutionService.java [hunk 4]
```diff
@@ -43,7 +43,7 @@ public PipelineExecutionService(PipelineStore store, ThreadPool threadPool) {
         this.threadPool = threadPool;
     }
 
-    public void execute(IndexRequest indexRequest, String pipelineId, ActionListener<IngestDocument> listener) {
+    public void execute(IndexRequest indexRequest, String pipelineId, ActionListener<Void> listener) {
```
Comment (martijnvg) [inline plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineExecutionService.java]: I'm not a big fan of ActionListener<Void>? Maybe we can do this differently and replace it with two functions? Runnable for the onResponse() part and for onFailure use a Consumer.

Comment (martijnvg) [inline plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineExecutionService.java]: but this can also be done in another pr.

File: plugins/ingest/src/main/java/org/elasticsearch/ingest/IngestDocument.java (hunk 2 above)
Comment (martijnvg) [inline plugins/ingest/src/main/java/org/elasticsearch/ingest/IngestDocument.java]: I think is fine as it is right now. I thought we weren't referencing to fields outside the class.

File: plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineExecutionService.java (hunk 4 above)
Comment (javanna) [inline plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineExecutionService.java]: sounds good to me, not a big fan of my change either. I just wanted to make sure that nobody relies on IngestDocument though here, cause it gets modified before the document is set back to the IndexRequest.

File: plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineExecutionService.java [hunk 5]
```diff
@@ -62,17 +62,16 @@ public void execute(IndexRequest indexRequest, String pipelineId, ActionListener
...
-                }
-                indexRequest.index(ingestDocument.getMetadata(IngestDocument.MetaData.INDEX));
-                indexRequest.type(ingestDocument.getMetadata(IngestDocument.MetaData.TYPE));
-                indexRequest.id(ingestDocument.getMetadata(IngestDocument.MetaData.ID));
-                indexRequest.routing(ingestDocument.getMetadata(IngestDocument.MetaData.ROUTING));
-                indexRequest.parent(ingestDocument.getMetadata(IngestDocument.MetaData.PARENT));
-                indexRequest.timestamp(ingestDocument.getMetadata(IngestDocument.MetaData.TIMESTAMP));
-                indexRequest.ttl(ingestDocument.getMetadata(IngestDocument.MetaData.TTL));
-                listener.onResponse(ingestDocument);
+                Map<IngestDocument.MetaData, String> metadataMap = ingestDocument.extractMetadata();
```
Comment (martijnvg) [inline plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineExecutionService.java]: maybe add a comment here, that it is ok to overwrite the meta data properties, because if they were set then they were copied when the ingest document was created?

File: plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineExecutionService.java (hunk 4 above)
Comment (martijnvg) [inline plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineExecutionService.java]: that is a good reason to remove the ingest document itself from the action listener.

**Comment by martijnvg** [comment]
LGTM
